samples/siwx91x_ota/src/main.c

```diff
@@ -0,0 +1,1043 @@
+/*
+ * @brief Prints the current firmware version
+ *
+ * Retrieves and displays the current firmware version information
+ * from the device.
+ */
+void ota_print_firmware_version(void)
+{
+	int ret = 0;
+	sl_wifi_firmware_version_t version = {};
+
+	/* Get initial firmware version */
+	ret = sl_wifi_get_firmware_version(&version);
+	if (ret == SL_STATUS_OK) {
+		print_firmware_version(&version);
+	} else {
+		printf("Failed to get firmware version: 0x%x\n", ret);
+	}
+}
+
+/*
+ * @brief Clean up resources used by the OTA update process
+ *
+ * @param app_ctx_st Pointer to application context structure containing resources to clean up
+ */
+void ota_cleanup_resources(struct app_ctx *app_ctx_st)
+{
+	if (app_ctx_st->sock >= 0) {
+		zsock_close(app_ctx_st->sock);
+		app_ctx_st->sock = -1;
+	}
+	k_msgq_purge(&app_ctx_st->http_data_q);
+	app_ctx_st->http_response_error = false;
+	app_ctx_st->ipv6_addr_config_done = false;
+}
+
+/*
+ * @brief Handles retry logic for OTA operations
+ *
+ * Manages retry attempts for various OTA operations, with appropriate
+ * cleanup between attempts.
+ *
+ * @param app_ctx_st Pointer to application context
+ * @param operation_name Name of the operation being retried for logging
+ * @param cleanup_socket Whether to clean up socket resources before retrying
+ * @return true if retry should continue, false if max retries exceeded
+ */
+static bool handle_retry(struct app_ctx *app_ctx_st, const char *operation_name,
+			 bool cleanup_socket)
+{
+	static uint8_t retry_count;
+
+	/* Clean up socket and message queue if needed */
+	if (cleanup_socket && app_ctx_st->sock >= 0) {
+		ota_cleanup_resources(app_ctx_st);
+	}
+	if (++retry_count > MAX_RETRIES) {
+		printf("Maximum retries exceeded, aborting OTA\n");
+		retry_count = 0;
+		return false;
+	}
+	printf("Retrying %s (%d/%d)...\n", operation_name, retry_count, MAX_RETRIES);
+	k_sleep(K_MSEC(1000));
+	return true;
+}
+
+/*
+ * @brief Main OTA application state machine
+ * @param app_ctx_st Pointer to application context
+ */
+static void ota_application_start(struct app_ctx *app_ctx_st)
+{
+	int ret = 0;
+	int firmware_status = 0;
+	char *ssid = CONFIG_OTA_WIFI_SSID;
+	char *pwd = CONFIG_OTA_WIFI_PSK;
+	bool ota_complete = false;
+
+	printf("OTA Application Started\n");
+	ota_print_firmware_version();
+
+	/* Main state machine loop */
+	while (1) {
+		switch (app_ctx_st->state) {
+		case OTA_STATE_SCAN:
+			ret = wifi_start_scan(app_ctx_st);
+			break;
+		case OTA_STATE_CONNECT:
+			ret = wifi_connect(ssid, pwd, CONFIG_OTA_WIFI_SECURITY_TYPE, app_ctx_st);
+			if (ret < 0) {
+				if (!handle_retry(app_ctx_st, "Wi-Fi connection", false)) {
+					return;
+				}
+				break;
+			}
+			break;
+
+		case OTA_STATE_IP_CONFIG:
+			ret = configure_ip(app_ctx_st);
+			if (ret < 0) {
+				if (!handle_retry(app_ctx_st, "IP configuration", false)) {
+					return;
+				}
+				break;
+			}
+			app_ctx_st->state = OTA_STATE_SERVER_CONNECT;
+			break;
+
+		case OTA_STATE_SERVER_CONNECT:
```

Comment:
`OTA_STATE_IP_CONFIG` can only transition to `OTA_STATE_SERVER_CONNECT`. So, they are in fact the same state.